surefire-providers/surefire-testng/src/main/java/org/apache/maven/surefire/testng/conf/TestNG740Configurator.java

```diff
@@ -37,33 +36,59 @@
  *
  * @since 3.0.0-M6
  */
-public class TestNG740Configurator extends TestNG60Configurator
+public class TestNG740Configurator
+    extends TestNG60Configurator
 {
     @Override
-    public void configure( XmlSuite suite, Map<String, String> options )
+    protected void configureThreadCount( XmlSuite suite, Map<String, String> options )
         throws TestSetFailedException
     {
-        String threadCountAsString = options.get( THREADCOUNT_PROP );
-        int threadCount = threadCountAsString == null ? 1 : parseInt( threadCountAsString );
-        suite.setThreadCount( threadCount );
-
-        String parallel = options.get( PARALLEL_PROP );
-        if ( parallel != null )
+        String threadCount = options.get( THREADCOUNT_PROP );
+        // if [threadcount] spec'd
+        if ( threadCount != null )
         {
-            if ( !"methods".equalsIgnoreCase( parallel ) && !"classes".equalsIgnoreCase( parallel ) )
+            try
             {
-                throw new TestSetFailedException( "Unsupported TestNG parallel setting: "
-                    + parallel + " ( only METHODS or CLASSES supported )" );
+                // convert and apply [threadcount] setting
+                suite.setThreadCount( Integer.parseInt( threadCount ) );
+            }
+            catch ( NumberFormatException e )
+            {
+                throw new TestSetFailedException( "Non-integer TestNG [threadcount] setting: " + threadCount, e );
             }
-            Class enumClass = tryLoadClass( XmlSuite.class.getClassLoader(), "org.testng.xml.XmlSuite$ParallelMode" );
-            Enum<?> parallelEnum = Enum.valueOf( enumClass, parallel.toUpperCase() );
-            invokeSetter( suite, "setParallel", enumClass, parallelEnum );
         }
-
-        String dataProviderThreadCount = options.get( "dataproviderthreadcount" );
-        if ( dataProviderThreadCount != null )
+    }
+    
+    @Override
+    protected void configureParallel( XmlSuite suite, Map<String, String> options )
+        throws TestSetFailedException
+    {
+        String parallel = options.get( PARALLEL_PROP );
+        // if [parallel] spec'd
+        if ( parallel != null )
         {
-            suite.setDataProviderThreadCount( Integer.parseInt( dataProviderThreadCount ) );
+            // try to load the [ParallelMode] enumeration
+            Class enumClass = tryLoadClass( XmlSuite.class.getClassLoader(), "org.testng.xml.XmlSuite$ParallelMode" );
+            // if enumeration loaded
+            if ( enumClass != null )
+            {
+                try
+                {
+                    // convert [parallel] option to corresponding constant
+                    Enum<?> parallelEnum = Enum.valueOf( enumClass, parallel.toUpperCase() );
+                    // set [XmlSuite] parallel mode to specified value
+                    invokeSetter( suite, "setParallel", enumClass, parallelEnum );
```

Comment:
Pleas add in comments version of TestNG which contains such enum and method `setParallel(Enum ...)`
We want to drop some of old TestNG, so maybe reflection would be removed.

Reply:
We are in the right version with this class. We do not expect null `enumClass`. The null check `if ( enumClass != null )` is hiding a future bug - we would not notice the bug. If we get NPE, we would immediately see it.

Reply:
I guess `org.testng.xml.XmlSuite$ParallelMode` should be FQN `org.testng.xml.XmlSuite.ParallelMode`.

Reply:
@slawekjaranowski: The class header JavaDoc specifies the TestNG version (`7.4.0`). I suppose this comment could be a bit more explicit, however, regarding the necessity for this configurator class.
@Tibor17: The inability to load the enumeration will trigger a clearly-explained failure: `Failed loading TestNG [ParallelMode] enumeration to convert [parallel] setting`

Reply:
@Tibor17: Regarding FQN, the **ParallelMode** enumeration is a nested class, so the name is properly spelt. However, your suggestion would be correct for an `import` statement. We would need to have the **XmlSuite** class definition from TestNG 7.4 for that to work, though.

Reply:
@slawekjaranowski: I added JavaDoc to the configuration methods. The documentation for the `configureParallel` method explains the rationale behind this method, specifying the TestNG release that made this necessary.